## Aim:
How to link the rule-based extracted CI_TYPE-GEO pairs with the respective Ci failure impacts

**Idea**\
Test using prompt engineering by passing table of CIGEO pairs to GPT-J model.
Steps:
* Load model and apply it always on one chunk of the document to extract CI failure impacts 
* Use prompt engineering to extract time and location of the CI failure (origin) the CI impacts (impact location)
or 
* Pass dataframe of pairs as input to the model
or
* Use few shot prompting with example answers

**Finally:**
* Compaire all approaches of spatial and temporal linking CI failure impacts

In [1]:
import os
import sys
import subprocess
import importlib

import re
import time
from glob import glob
from pathlib import Path

from io import StringIO

import numpy as np
import pandas as pd
from jinja2 import Environment, FileSystemLoader
from langchain_docling import DoclingLoader
import langdetect

import spacy
from huggingface_hub import login
import transformers
from transformers import (
    AutoTokenizer,
    AutoModelForCausalLM,
    AutoModelForSeq2SeqLM,
    BitsAndBytesConfig,
)

import torch

from pdfminer.high_level import extract_text
from docling.backend.pypdfium2_backend import PyPdfiumDocumentBackend
from docling.datamodel.base_models import InputFormat
from docling.datamodel.pipeline_options import (
    PdfPipelineOptions,
    AcceleratorOptions,
    AcceleratorDevice,
)
from docling.pipeline.standard_pdf_pipeline import StandardPdfPipeline
from docling.document_converter import DocumentConverter, FormatOption


sys.path.append("../")
from src.settings import settings as s
import src.document_cleaning as dc
import src.translation_model as tm
# from src.hf_model_access import get_weight_dir


torch.manual_seed(42)

#  automatic linebreaks and multi-line cells.
pd.set_option("display.max_colwidth", None)
pd.set_option("display.colheader_justify", "left")


test_mode = True

torch.cuda.empty_cache() 
print(torch.cuda.memory_reserved() / 1e9)


Running on local machine
0.0


In [2]:
# os.environ["HF_HOME"] = s.HF_HOME_DIR
try: 
    login(token=os.getenv("HUGGINGFACE_TOKEN"))   # notebook_login
except:
    login(token=os.environ.get("HUGGINGFACE_TOKEN"))  # former HF_TOKEN


test_mode = True # for testing purposes, limit number of processed documents



Note: Environment variable`HF_TOKEN` is set and is the current active token independently from the token you've just configured.


## Set paths and vars

In [3]:
# set wd to project root
os.chdir("/home/a-buch/Documents/TUB_DWN/_PROJECTS/CI-impacts-information-retrieval")



## set path variables
DOCS_DIR = Path(s.PATH_DATA + "text_sources/")
PARSED_TEXT_DIR = Path(s.PATH_DATA + "parsed_documents/")
LLM_OUTPUTS_DIR = Path(s.PATH_DATA + "llm_outputs/")
NER_PATTERNS_FILEPATH = Path(s.NER_PATTERNS_FILEPATH)

# document cleaning
md_dir = Path(PARSED_TEXT_DIR)
md_dir.mkdir(parents=True, exist_ok=True)

# CI GEO pairs
CI_GEO_FILEPATH = Path("./" + s.PATH_DATA + s.CI_GEO_PAIRS_FILENAME)

## store LLM 1 response and prompt
os.makedirs(s.PATH_LLM_DATA, exist_ok=True)
OUTPUT_LLM1_FILEPATH =  Path(s.PATH_LLM_DATA / s.LLM_DATA_FILENAME)
OUTPUT_PROMPT_FILEPATH = Path(s.PATH_LLM_DATA / f"prompt_{OUTPUT_LLM1_FILEPATH.stem}.txt" )




### Set test mode

In [4]:

if test_mode:
    print("Test mode is ON. Using only a small sample of documents for testing.")

    docs_list_sample = [
        Path(PARSED_TEXT_DIR, "Karakatsani 2023 - Greece economy briefing The economic impact of the recent devastating floods in Greece_cleaned.md"),
        Path(PARSED_TEXT_DIR, "Koks 2022 - Brief communication_cleaned.md"),
        Path(PARSED_TEXT_DIR, "Khazai 2013 - Juni-Hochwasser 2013 in Mitteleuropa - Fokus Deutschland Bericht 2 Auswirkungen und Bewältigung_cleaned.md"),
        Path(PARSED_TEXT_DIR, "European Investment Bank 2025 - Spain_ EIB lends €50 million to Iberdrola to rebuild and climate-proof flood-hit power infrastructure in Valencia_cleaned.md"),
        Path(PARSED_TEXT_DIR, "Wilson 2024 - Flash floods in Spain sweep away cars, disrupt trains and leave several missing _ AP News_cleaned.md"),     
        Path(PARSED_TEXT_DIR, "Wildhagen 2013 - Hochwasser_ Wie die Flut Unternehmen lahmlegt_cleaned.md"),
        Path(PARSED_TEXT_DIR, "Lloyd's List 2024 - Port of Valencia reopens after devastating floods_cleaned.md"),
        Path(PARSED_TEXT_DIR, "Containerlift 2024 - Valencia Port Resumes Operations Following Devastating Flooding in Spain - Containerlift.co.uk - Transport_Lifting_Shipping_cleaned.md"), 
        Path(PARSED_TEXT_DIR, "AEMET 2024 - ESTUDIO SOBRE LA SITUACIÓN DE LLUVIAS INTENSAS_cleaned.md")
    ]


## Test mode
if test_mode:
    search_path = docs_list_sample
    print("Test mode is ON. Using only a small sample of documents for testing.")
else:
    search_path = glob(str(Path(PARSED_TEXT_DIR, "*cleaned.md")))


Test mode is ON. Using only a small sample of documents for testing.
Test mode is ON. Using only a small sample of documents for testing.


###  Load and initialize spaCy language model

In [5]:
# # spaCy language model for NER and ENTITY LINKING

try: 
    print("Try loading spaCy language model for local machine ...")
    try:
        nlp = spacy.load(s.SPACY_MODEL)
    except (OSError, ValueError):
        print(f"spaCy language model '{s.SPACY_MODEL}' not found. Downloading ...")
        ## loading transformer language model for NER requires additional package
        if (s.SPACY_MODEL.endswith("_trf")):
            !uv add spacy[transformers]
        !uv run python -m spacy download {s.SPACY_MODEL}
        nlp = spacy.load(s.SPACY_MODEL)

except (OSError, ValueError):
    try: 
        print(f"spaCy language model '{s.SPACY_MODEL}' not found. Downloading ...")
        subprocess.check_call(["uv", "pip", "install", "spacy-transformers"])
        subprocess.check_call(["uv", "run", "python3", "-m", "spacy", "download", s.SPACY_MODEL])
        nlp = spacy.load(s.SPACY_MODEL)
    except ValueError:
        print("Try downloading directly via python3 -m spacy ... (and not via uv run ...)")
        subprocess.check_call(["python3", "-m", "spacy", "download", s.SPACY_MODEL])


# !uv run python -m spacy download en_core_web_trf
# nlp = spacy.load("en_core_web_trf")

Try loading spaCy language model for local machine ...


In [6]:
# Initialise spaCy pipeline 
nlp.add_pipe("merge_entities")
nlp.add_pipe("merge_noun_chunks")


# add CI_TYPE patterns to spacy nlp model pipeline
config = {"spans_key": None, "annotate_ents": True, "overwrite": False}
## see for more info: https://spacy.io/usage/rule-based-matching#entityruler
## NOTE EntityRuler is hidden inside .add_pipe()
try:
    ruler = nlp.add_pipe("span_ruler", config=config)
    ruler.from_disk(s.NER_PATTERNS_FILEPATH)
except ValueError:
    print("SpanRuler already exists in pipeline.")
    ruler = nlp.get_pipe("span_ruler")
    ruler.from_disk(s.NER_PATTERNS_FILEPATH)


In [7]:
## Document cleaning

In [7]:
# Docling pipeline configs

accelerator_options = AcceleratorOptions(
    num_threads=4, device=AcceleratorDevice.AUTO
)  # use GPU + multi-threading
pipeline_options = PdfPipelineOptions()
pipeline_options.do_ocr = True
pipeline_options.do_table_structure = (
    True  # identify tables as such just not to have them in the TextItems later
)
pipeline_options.accelerator_options = accelerator_options
pipeline_options.force_backend_text = True




# setup converter for PDF and markdown
converted = DocumentConverter(
    allowed_formats=[InputFormat.PDF, InputFormat.MD],
    format_options={
        InputFormat.PDF: FormatOption(
            pipeline_cls=StandardPdfPipeline,
            pipeline_options=pipeline_options,
            backend=PyPdfiumDocumentBackend,
        ),
    },
)


In [8]:
print( "Number of documents to process:", len(os.listdir(DOCS_DIR)) )


# convert the different layouts of the pdf files into unified markdown format incl. sub/section titles, tables, caption text etc
for pdf_filename in os.listdir(DOCS_DIR):
    if pdf_filename.endswith(".pdf"):

        md_filename = f"{Path(pdf_filename).stem}.md"

        pdf_filepath = os.path.join(DOCS_DIR, Path(pdf_filename))
        md_filepath = os.path.join(PARSED_TEXT_DIR, Path(md_filename))
        cleaned_md_filepath = md_filepath.replace(".md", "_cleaned.md")

        if os.path.exists(md_filepath):
            print(
                f"Markdown file '{md_filepath}' already exists. Skipping conversion and cleaning."
            )
            continue

        start_time = time.time()
        print(f"\nFetching: {pdf_filename}")

        print("Remove reference section")
        pdf_text = extract_text(pdf_filepath)
        pdf_text_no_refs = dc.remove_references(pdf_text)

        print("Removing URLs") # LangExtract tries to open these URLs when they occur in the document text
        pdf_text_no_refs_urls = re.sub(r"http\S+", "", pdf_text_no_refs) 

        # FIXME remove workaround of saving pdf as markdown and reading it again as Docling.Document
        # FIXME with DocLoader
        # loader = DoclingLoader(md_filepath)
        # md_text = loader.load()
        with open(md_filepath, "w", encoding="utf-8") as f:
            f.write(pdf_text_no_refs_urls)
        print("Converting Markdown to text...")
        md_text = converted.convert(md_filepath)

        print("Removing headers and footers\n")
        md_text_cleaned = dc.remove_headers_footers(md_text)

        print(f"Saving parsed and cleaned document as markdown to: {cleaned_md_filepath}")
        md_text_cleaned.document.save_as_markdown(cleaned_md_filepath)

        end_time = time.time() - start_time
        print(f"Parsing and cleaning done. Time elapsed: {end_time:.2f} seconds.")


# visual check of removed items
# TODO make as document_cleaning function: print removed items with largest number of chars first
# ## NOTE. high number of chars == more potentially actual text body

# text_items_removed = sorted(text_items_to_drop_visualization, key=lambda x: -x[0])
# for i in text_items_removed[:50]:
#     print(i) # -->  also subsection titles were removed partly





Number of documents to process: 47
Markdown file '../data/parsed_documents/UNEP 2013 - Impacts of summer 2003.md' already exists. Skipping conversion and cleaning.
Markdown file '../data/parsed_documents/Stamataki 2023 - Greece’s record rainfall and flash floods are part of a trend _ PreventionWeb.md' already exists. Skipping conversion and cleaning.
Markdown file '../data/parsed_documents/Khazai 2013 - Juni-Hochwasser 2013 in Mitteleuropa - Fokus Deutschland Bericht 2 Auswirkungen und Bewältigung.md' already exists. Skipping conversion and cleaning.
Markdown file '../data/parsed_documents/Koks 2022 - Brief communication.md' already exists. Skipping conversion and cleaning.
Markdown file '../data/parsed_documents/Yamashita 2008 - Analysis  control  and economic impact assessment of major blackout events.md' already exists. Skipping conversion and cleaning.
Markdown file '../data/parsed_documents/AEMET 2024 - ESTUDIO SOBRE LA SITUACIÓN DE LLUVIAS INTENSAS.md' already exists. Skipping co

## CI GEO PAIRS

In [ ]:
# ## see for more info: https://spacy.io/usage/rule-based-matching#entityruler
# ## NOTE EntityRuler is hidden inside .add_pipe()


# ## call nlp model and create pipeline with new entity pattern
# # NOTE Creation of the new entity (CI_TYPE) solves the issue that FAC entities 
# # (buildings, airports, highways, bridges, etc.) refer only to the name of the facility (e.g. A76, Ahrtalbahn)
# config = {"spans_key": None, "annotate_ents": True, "overwrite": False}
# try:
#     ruler = nlp.add_pipe("span_ruler", config=config)
#     ruler.from_disk(s.NER_PATTERNS_FILEPATH)
# except ValueError:
#     print("SpanRuler already exists in pipeline.")
#     ruler = nlp.get_pipe("span_ruler")
#     ruler.from_disk(s.NER_PATTERNS_FILEPATH)



In [ ]:
# test_mode = True

# if test_mode:
#     print("Test mode is ON. Using only a small sample of documents for testing.")
    
#     docs_list_sample = [
#         # Path(PARSED_TEXT_DIR, "Karakatsani 2023 - Greece economy briefing The economic impact of the recent devastating floods in Greece_cleaned.md"),
#         # Path(PARSED_TEXT_DIR, "Koks 2022 - Brief communication_cleaned.md"),
        
#         # Path(PARSED_TEXT_DIR, "Wildhagen 2013 - Hochwasser_ Wie die Flut Unternehmen lahmlegt_cleaned.md"),
#         # Path(PARSED_TEXT_DIR, "Khazai 2013 - Juni-Hochwasser 2013 in Mitteleuropa - Fokus Deutschland Bericht 2 Auswirkungen und Bewältigung_cleaned.md"),
#         Path(PARSED_TEXT_DIR, "European Investment Bank 2025 - Spain_ EIB lends €50 million to Iberdrola to rebuild and climate-proof flood-hit power infrastructure in Valencia_cleaned.md"),
#         Path(PARSED_TEXT_DIR, "Wilson 2024 - Flash floods in Spain sweep away cars, disrupt trains and leave several missing _ AP News_cleaned.md"),     
#         Path(PARSED_TEXT_DIR, "Lloyd's List 2024 - Port of Valencia reopens after devastating floods_cleaned.md"),
#         Path(PARSED_TEXT_DIR, "Containerlift 2024 - Valencia Port Resumes Operations Following Devastating Flooding in Spain - Containerlift.co.uk - Transport_Lifting_Shipping_cleaned.md"), 
#         # Path(PARSED_TEXT_DIR, "AEMET 2024 - ESTUDIO SOBRE LA SITUACIÓN DE LLUVIAS INTENSAS_cleaned.md")
#     ]



In [ ]:
# # token.head.lefts[0]#.dep_ #.lefts[0]
# for w in token.head.rights:
#     print(token.head, w, w.dep_, w.pos_)


In [29]:
TEXTS = [
    "A major storm briefly closed down the gas supply in New York City, and the water network and wastewater supply in Jersey City", 
]
doc = nlp( TEXTS[0])
spacy.displacy.render(doc, style="dep", jupyter=True)


In [ ]:
# subtrees: get a whole phrase by its syntactic head using the Token.subtree
loader = DoclingLoader(filepath)
doc = loader.load()
nlp = spacy.load("en_core_web_trf")
#doc = nlp("Credit and mortgage account holders must submit their requests")
doc = nlp( doc[13].page_content)

for i in range(0, 7):
    root = [token for token in doc if token.head == token][i]
    subject = list(root.lefts)[0]
    print("\nRoot:", root)
    print("Root lefts:", list(root.lefts))
    print("Root rights:", list(root.rights))
    for descendant in subject.subtree:
        # assert subject is descendant or subject.is_ancestor(descendant) # only subtrees
        assert subject is descendant or subject.is_ancestor(descendant) or subject == descendant
        print(descendant.text, descendant.dep_, descendant.n_lefts, descendant.n_rights,
                [ancestor.text for ancestor in descendant.ancestors])

displacy.render(doc, style="dep", jupyter=True)


In [ ]:

## overwirte previous spacy loading for testing 
nlp = spacy.load("en_core_web_trf")

nlp.add_pipe("merge_entities")
nlp.add_pipe("merge_noun_chunks")

config = {"spans_key": None, "annotate_ents": True, "overwrite": False}
try:
    ruler = nlp.add_pipe("span_ruler", config=config)
    ruler.from_disk(s.NER_PATTERNS_FILEPATH)
except ValueError:
    print("SpanRuler already exists in pipeline.")
    ruler = nlp.get_pipe("span_ruler")
    ruler.from_disk(s.NER_PATTERNS_FILEPATH)


TEXTS = [
    "The water supply fall out. This happened in the city of Bad Münstereifel",
    "The wastewater supply fall out. This happened in the city of Bad Münstereifel",
    "Water supply fall out. This happened in the city of Bad Münstereifel",
    # doc[13].page_content,
    "In the heavily destroyed town of Bad Münstereifel (in the state of North Rhine-Westphalia), water supply was re-established within 5 d after the flood event",
    
    # cant find CI_TYPE entity "water supply" in this sentence
    "In Bad Münstereifel the water supply was affected. In the same region the Ahrtalbahn was cancelled as well. "
]

for d in nlp.pipe(TEXTS):
    print(d)
    for token in d:
        if token.ent_type_ in ["CI_TYPE", "FAC", "GPE", "LOC"]:
            print("\ntoken: ", token, token.dep_, token.pos_, token.ent_type_)



In [ ]:

nlp = spacy.load("en_core_web_trf")
doc = nlp("Credit and mortgage account holders must submit their requests")
span = doc[doc[4].left_edge.i : doc[4].right_edge.i+1]
with doc.retokenize() as retokenizer:
    retokenizer.merge(span)
for token in doc:
    print(token.text, token.pos_, token.dep_, token.head.text)


In [ ]:
# [ancestor.text for ancestor in descendant_loc.ancestors]
# for i in sentence_root.subtree :
#     print(i)
    # for i in i.children:
    #     print("  ", i)



In [ ]:
#  Setup spaCy  pipeline 
nlp = spacy.load("en_core_web_trf")
nlp.add_pipe("merge_entities")
nlp.add_pipe("merge_noun_chunks")

# add CI_TYPE to nlp model pipeline
config = {"spans_key": None, "annotate_ents": True, "overwrite": False}
try:
    ruler = nlp.add_pipe("span_ruler", config=config)
    ruler.from_disk(s.NER_PATTERNS_FILEPATH)
except ValueError:
    print("SpanRuler already exists in pipeline.")
    ruler = nlp.get_pipe("span_ruler")
    ruler.from_disk(s.NER_PATTERNS_FILEPATH)


# run trough text per chunk
TEXTS = [
    #"A major storm briefly closed down the gas supply in New York City, and the water network and wastewater supply in Jersey City",
    Path(PARSED_TEXT_DIR, "Karakatsani 2023 - Greece economy briefing The economic impact of the recent devastating floods in Greece_cleaned.md"),
    #Path(PARSED_TEXT_DIR, "Koks 2022 - Brief communication_cleaned.md"),
    # doc[13].page_content,
]


for filepath in TEXTS:

    loader = DoclingLoader(filepath)
    doc = loader.load()

    for chunk in doc:
        # applied NLP pipeline -> adding CI_entity annotations
        for nlp_chunk in nlp.pipe([chunk.page_content]):

            ## NOTE: FAC leads also of extraction of non CI, like national park, sprot park etc. 
            ## thus remove FAC and run only with CI_TYPE
            ci_entities = [ent for ent in nlp_chunk.ents if ent.label_ in ["CI_TYPE"]]
            # ci_entities = [ent for ent in nlp_chunk.ents if ent.label_ in ["CI_TYPE", "FAC"]] #, "GPE", "LOC"]]

            for ci_ent in ci_entities:
                print("ci_ent:", ci_ent.text)

                # Get the root of the sentence to get all CI-GEO pairs of the sentence
                # head = ci_ent.root.head  # only direct head of ci_entity, but not the sentence root
                sentence_root = ci_ent.sent.root

                if sentence_root.dep_ in ("ROOT", "dobj", "conj"):  # TODO probably enough when set only to ROOT 
                #if sentence_root.pos_ == "VERB" and sentence_root.dep_ == "ROOT":

                    head = ci_ent.root.head
                    
                    # print(f" `{sentence_root, sentence_root.dep_}` is sentence_root, \n `{head}` the direct head of ci entity `{ci_ent}`")
                    
                    # preps = [token for token in head.children if token.dep_ in ("prep", "pobj", "nsubjpass")]
                    # for prep in preps:
                    # "                #     # Check if tokens part of LOC entities are in the preposition's (prep=at)
                    # "                #     # children, e.g. at -> Acme Corp Inc. , same. in --> Bad Münstereifel
                    # "                #     locs = [token for token in prep.children if token.ent_type_ in ("GPE", "LOC")]

                    subjs = list(sentence_root.lefts) + list(sentence_root.rights)
                    for subj in subjs:
                        for descendant in subj.subtree:

                            # check if descendant is a location entity
                            if descendant.ent_type_ in ("GPE", "LOC"):
                            # if descendant.ent_type_ in ("GPE", "LOC") and descendant.dep_ in ("prep", "pobj", "conj"): # "dobj", "nsubjpass"): 
                            # if descendant_loc.pos_ in ("NNP", "PROPN") and descendant_loc.dep_ in ("prep", "pobj", "conj"): # "dobj", "nsubjpass"): 
                                
                                descendant_loc = descendant
                                # print(
                                #     "locs descendant:", descendant_loc.text, 
                                #     "\nits ancestors", [ancestor.text for ancestor in descendant_loc.ancestors]
                                # )
                                print("  ", ci_ent.text, "->", descendant_loc.text)
                                # assert subj is descendant_loc or subj.is_ancestor(descendant_loc)
                                

                                ###  --- Special case handling - one root but multiple ci-geo pairs  --- ###
                                ###  --- Text example:  "A major storm briefly closed down the gas supply in New York City, and the water network and wastewater supply in Jersey City", 
                                
                                # FIXME : hack: checks only that CI_ENT is equal to first CI_TYPE element in ancestors, but skips later ancestors of CI_TYPE
                                # issue: descendant_loc (Jersey city) was linked to also later CI_TYPEs, 
                                #   issue text example:  "A major storm briefly closed down the gas supply in New York City, and the water network and wastewater supply in Jersey City", 
                                #   issue example: descendant of location (Jersey City) its ancestors: ['in', 'the water network', 'the gas supply', 'closed']
                                # solution: link descendant_loc to only the first CI_TYPE which is usually the correct one
                                #   BUT causes smaller issue that "wastewater supply" is not linked to "Jersey City",
                                #   because its head is "the water network "
                                
                                try:
                                    first_ci_ancestor = [ancestor.text for ancestor in descendant_loc.ancestors if ancestor.ent_type_ in ("CI_TYPE", "FAC")][0]
                                    if ci_ent.text == first_ci_ancestor:
                                        print(
                                            "descendant of location:", descendant_loc.text, descendant_loc.dep_, 
                                            "\nits ancestors:", [ancestor.text for ancestor in descendant_loc.ancestors] # all parents up to root.head
                                        )
                                        print("  ", ci_ent.text, "-->", descendant_loc.text)
                                except IndexError:
                                    pass



# spacy.displacy.render(d, style="dep", jupyter=True)
# token.dep_ == "pobj" or token.dep_ == "dobj" and token.head.dep_ == "prep":

#### Apply new CI-Geo extraction on entire documents

In [ ]:
# Initalize spaCy pipeline 
nlp = spacy.load(s.SPACY_MODEL)
nlp.add_pipe("merge_entities")
nlp.add_pipe("merge_noun_chunks")


# add CI_TYPE to nlp model pipeline
config = {"spans_key": None, "annotate_ents": True, "overwrite": False}
try:
    ruler = nlp.add_pipe("span_ruler", config=config)
    ruler.from_disk(s.NER_PATTERNS_FILEPATH)
except ValueError:
    print("SpanRuler already exists in pipeline.")
    ruler = nlp.get_pipe("span_ruler")
    ruler.from_disk(s.NER_PATTERNS_FILEPATH)


if test_mode:
    search_path = docs_list_sample
    print("Test mode is ON. Using only a small sample of documents for testing.")
else:
    search_path = glob(str(Path(PARSED_TEXT_DIR, "*cleaned.md")))




In [ ]:



## TODO make as pydantic class with fixed attributes
df_ci_geo = pd.DataFrame(
    columns=[
        "citation_id",
        "chunk_id",
        "ci_entity",
        "geo_entity",
        "chunk text",
    ]
)

sentence_root_list = [] # for visualization of sentence roots and their dependencies
time0 = time.time()

## Start CI impact extraction
for i, filename in enumerate(search_path):

    no_documents = len(search_path)
    filepath = Path(filename)
    filename_stem = filepath.stem


    ## load chunks
    loader = DoclingLoader(filepath)
    doc = loader.load()



    print(f"\n\n ######## -------- Processing document [{i+1}/{no_documents}]: {filepath.name} -------- ######## \n")

    ## extract authors, publication year and title 
    author, year, title = dc.extract_citation_info(filename_stem)
    citation = f"{author} {year}".replace("  ", " ").strip()
    title = title.replace(" - ", "").replace("_cleaned", "").strip()


    print(f"\n ######## -------- Relationship extraction: CI-GEO pairs -------- ######## \n")


    ## get most likely geolocation for each CI entity based on distance
    for chunk_no, chunk in enumerate(doc):

        ## apply customized NLP pipeline -> returns same chunk.text but with CI_entities annotations and without linebreak symbols
        for nlp_chunk in nlp.pipe([chunk.page_content]):


            ## NOTE: FAC leads also of extraction of non CI, like national park, sprot park etc. 
            ## thus remove FAC and run only with CI_TYPE
            ci_entities = [ent for ent in nlp_chunk.ents if ent.label_ in ["CI_TYPE"]]
            # ci_type_ents = [ent for ent in nlp_chunk.ents if ent.label_ in ["CI_TYPE", "FAC"]]

            # check if chunk contains CI_TYPE entities
            if not ci_entities:
                print(f"Chunk [{chunk_no}], No CI_TYPE entities found {len(ci_entities)}")
                # print(f"\nChunk [{chunk_no}], No. CI_TYPE and FAC entities: {len(ci_entities)}")
                continue

            # # iterate over all entities within chunk
            # for ent_idx in range(len(all_ents)):

            #     # when entity is CI_TYPE or FAC (i.e. buidling, airports, highways) do following ...
            #     if all_ents[ent_idx].label_ in ["CI_TYPE", "FAC"]:
            #         ci_idx = ent_idx
            
            print(f"Chunk [{chunk_no}], {len(ci_entities)} CI_TYPE  entities found")
            
            for ci_ent in ci_entities:
                print("Ci_ent:", ci_ent.text)

                # Get the root of the sentence 
                sentence_root = ci_ent.sent.root

                if sentence_root.dep_ in ("ROOT", "dobj", "conj"):  # TODO probably enough when set only to ROOT 
                    
                    sentence_root_list.append(sentence_root.dep_) # for visualization of sentence roots and their dependencies

                    head = ci_ent.root.head
                    # print(f" `{sentence_root, sentence_root.dep_}` is sentence_root, \n `{head}` the direct head of ci entity `{ci_ent}`")

                    subjs = list(sentence_root.lefts) + list(sentence_root.rights)
                    for subj in subjs:
                        for descendant in subj.subtree:

                            # check if descendant is a location entity
                            if descendant.ent_type_ in ("GPE", "LOC"):
                            # if descendant.ent_type_ in ("GPE", "LOC") and descendant.dep_ in ("prep", "pobj", "conj"): # "dobj", "nsubjpass"): 
                                
                                descendant_loc = descendant
                                
                                print("  ", ci_ent.text, "->", descendant_loc.text)
                                # assert subj is descendant_loc or subj.is_ancestor(descendant_loc)

                                ## Special case handling - one root but multiple ci-geo pairs
                                try:
                                    first_ci_ancestor = [ancestor.text for ancestor in descendant_loc.ancestors if ancestor.ent_type_ in ("CI_TYPE", "FAC")][0]
                                    if ci_ent.text == first_ci_ancestor:
                                        print(
                                            "descendant of location:", descendant_loc.text, descendant_loc.dep_, 
                                            "\nits ancestors:", [ancestor.text for ancestor in descendant_loc.ancestors] # all parents up to root.head
                                        )
                                        print("  ", ci_ent.text, "-->", descendant_loc.text)
                                        
                                        ## write as dict entry incl chunk_id, ci_entity, geo_entity, distance
                                        result_dict = {
                                            "citation_id": citation,
                                            "chunk_id": chunk_no,
                                            "ci_entity": ci_ent.text,
                                            "geo_entity": descendant_loc.text,
                                            "case_type": "special",
                                            "chunk text": nlp_chunk,
                                        }
                                        df_ci_geo = pd.concat(
                                            [df_ci_geo, pd.DataFrame([result_dict])], ignore_index=True
                                        )                                        

                                ## usual case handling
                                except IndexError:
                                    ## write as dict entry incl chunk_id, ci_entity, geo_entity, distance
                                    result_dict = {
                                        "citation_id": citation,
                                        "chunk_id": chunk_no,
                                        "ci_entity": ci_ent.text,
                                        "geo_entity": descendant_loc.text,
                                        "case_type": "usual",
                                        "chunk text": nlp_chunk,
                                    }
                                    df_ci_geo = pd.concat(
                                        [df_ci_geo, pd.DataFrame([result_dict])], ignore_index=True
                                    )


                else:
                    # print("\nNo CI_TYPE or FAC entities found in this chunk.")
                    continue

## post-process of DF CI-GEO pairs
df_ci_geo = df_ci_geo.drop_duplicates(
    subset=["citation_id", "chunk_id", "ci_entity", "geo_entity","case_type", "chunk text"]
    )# .reset_index(drop=True, inplace=True)

time1 = time.time() - time0

print(f"\nCI-GEO extraction done. Time elapsed: {time1/60:.2f} minutes.")


In [ ]:
set(sentence_root_list)

In [ ]:
## Fix issue that df_ci_geo now give multiple LOC for same CI_TYPE,   
dd = df_ci_geo.drop_duplicates(subset=["citation_id", "chunk_id", "ci_entity", "geo_entity", "chunk text"])# .reset_index(drop=True, inplace=True)
dd

In [ ]:
dd

In [ ]:
#### OLD CI-GEO CODE #############

# ## load docs
# docs_list = glob(PARSED_TEXT_DIR + "*_cleaned.md")
# print(f"Found {len(docs_list)} cleaned documents.")


# ## DataFrame to store CI-GEO entity pairs
# df_ci_geo = pd.DataFrame(     ## TODO make as pydantic class with fixed attributes
#     columns=[
#         "document_id",
#         "chunk_id",
#         "ci_entity",
#         "ci_entity_label",
#         "geo_entity",
#         "geo_entity_label",
#         "token_distance",
#     ]
# )


# ## iterate over all cleaned documents and extract CI-GEO entity pairs
# # but first check if output file already exists
# if OUTPUT_CI_GEOFILEPATH.exists():
#     print(f"\nCI-Geo entities already exists, see file:  {OUTPUT_CI_GEOFILEPATH.name}. \nLoading file from disk")

#     with open(OUTPUT_CI_GEOFILEPATH, 'r') as f:
#         df_ci_geo = pd.read_csv(f)

# else:
#     for FILE_PATH in docs_list:
#         print(f"\n\n Loading  - {Path(FILE_PATH).name} - ")
#         loader = DoclingLoader(FILE_PATH)  # use chunks from Docling.Loader
#         doc = loader.load()

#         ## get most likely geolocation for each CI entity based on distance
#         for i, chunk in enumerate(doc):
#             nlp_chunk = nlp(chunk.page_content)
#             all_ents = [ent for ent in nlp_chunk.ents]
#             ci_type_ents = [ent for ent in nlp_chunk.ents if ent.label_ in ["CI_TYPE", "FAC"]]
#             ci_type_ents_info = [ent for ent in nlp_chunk.ents if ent.label_ in ["CI_TYPE"]]
#             fac_ents_info = [ent for ent in nlp_chunk.ents if ent.label_ in ["FAC"]]

#             # check if chunk contains CI_TYPE entities
#             if len(ci_type_ents) > 0:
#                 print(f"\nChunk [{i}], No. CI_TYPE and FAC entities: {len(ci_type_ents)}")
#                 print(
#                     f"Contains following entities for CI_TYPE: {ci_type_ents_info}, FAC: {fac_ents_info}"
#                 )
#                 print(f"Chunk text [{i}]:", chunk.page_content)
#                 # print(f"{ {(ci_type_ents[i].text, ci_type_ents[i].label_) for i in range(len(ci_type_ents))} } ")

#                 # iterate over all entities within chunk
#                 for ent_idx in range(len(all_ents)):
#                     # when entity is CI_TYPE or FAC (i.e. buidling, airports, highways) do following ...
#                     if all_ents[ent_idx].label_ in ["CI_TYPE", "FAC"]:
#                         ci_idx = ent_idx

#                         ## .. calculate distances between CI_TYPE entity and  all GEO entities in chunk based on index position
#                         distance_list = []
#                         idx_in_chunk = []
#                         try:
#                             for ent_idx in range(len(all_ents)):
#                                 # TODO calc distances between CI_TYPE ~ GEO entities based on word numbers and not entities
#                                 if all_ents[ent_idx].label_ in ["GPE", "LOC"]:
#                                     geo_idx = ent_idx
#                                     dist_ent_pair = np.abs(ci_idx - geo_idx)
#                                     distance_list.append(dist_ent_pair)
#                                     idx_in_chunk.append((ent_idx))
#                                     closest_pair_idx = np.argmin(
#                                         distance_list
#                                     )  # idx of closest GEO entity
#                                     distance_closest_pair = distance_list[closest_pair_idx]

#                             threshold = 5  # max token distance between CI_TYPE and GEO entity
#                             if distance_closest_pair > threshold:
#                                 print(
#                                     f""" Token distance between CI_TYPE/FAR and next GEO entity is {distance_closest_pair} and thus larger than the allowed distance of {threshold} tokens """
#                                 )
#                                 continue
#                             else:
#                                 print(
#                                     f"""  Closest GEO entity to CI_TYPE/FAC entity "{all_ents[ci_idx]}" is "{all_ents[idx_in_chunk[closest_pair_idx]]}" at distance {distance_closest_pair}"""
#                                 )  # TODO constrain min.distance to max value (eg. 5 tokens), issue: likely when distance value is high that geolocation of Ci_type is mentioned in previous sentences or chunk

#                             ## write as dict entry incl chunk_id, ci_entity, geo_entity, distance
#                             result_dict = {
#                                 "document_id": Path(FILE_PATH).stem,
#                                 "chunk_id": i,
#                                 "ci_entity": all_ents[ci_idx].text,
#                                 "ci_entity_label": all_ents[ci_idx].label_,
#                                 "geo_entity": all_ents[idx_in_chunk[closest_pair_idx]].text,
#                                 "geo_entity_label": all_ents[
#                                     idx_in_chunk[closest_pair_idx]
#                                 ].label_,
#                                 "token_distance": distance_closest_pair,
#                             }
#                             df_ci_geo = pd.concat(
#                                 [df_ci_geo, pd.DataFrame([result_dict])], ignore_index=True
#                             )

#                         except IndexError:
#                             print("No GEO entities found in this chunk.")
#                             continue
#                         # print("\nidx_in_chunk, closest pair idx", idx_in_chunk, closest_pair_idx)

#                         # spacy.displacy.render(
#                         #     nlp_chunk, style="ent",
#                         #     options={"ents": ["CI_TYPE", "GPE", "LOC", "FAC"], "colors": {"CI_TYPE": "violet"}}
#                         # )

#                 else:
#                     print("\nNo CI_TYPE or FAC entities found in this chunk.")
#                     continue

#     # save to disk when not existing
#     with (OUTPUT_CI_GEOFILEPATH).open('w') as f:
#         df_ci_geo.to_csv(f, index=False)
#     print(f"\nSaved extracted CI-GEO entity pairs to {OUTPUT_CI_GEOFILEPATH}")

In [ ]:
df_ci_geo.loc[df_ci_geo["ci_entity"] == "rail"]  # .head(15)


## LLama

###  Prompt engineering

In [9]:
question_1 = "Which infrastructure failures are mentioned in the text? Categorize the output by the type of infrastructure, the location, and the type of damage."

In [10]:
env = Environment(loader=FileSystemLoader("./prompt_templates/"))
template = env.get_template("ci_loc_direct_impacts.txt")


### Init Llama application function 
* extract CI-GEO pairs (ie. most likely geolocation of each CI_TYPE )
* pass NER table to prompt for evaluating and improving LLM response
* Test approach by applying it on three cleaned documents


In [11]:
# # empty CUDA cache
import gc
import torch

gc.collect()

torch.cuda.empty_cache()
torch.no_grad()
# print(torch.cuda.memory_summary(device=None, abbreviated=False))



In [ ]:
# !uv add https://github.com/mjun0812/flash-attention-prebuild-wheels/releases/download/v0.4.22/flash_attn-2.8.1+cu130torch2.9-cp312-cp312-linux_x86_64.whl

Resolved 356 packages in 0.94ms                                      
Audited 352 packages in 3ms


In [14]:
# model_name = "meta-llama/Llama-2-7b-chat-hf"

# base_dir =  s.HF_HOME_DIR   # use default dir in .cache/
# model_dir = base_dir #+ f"/models--{model_name.replace("/", "--")}"  # is already .._mirror/hub/

# print(model_dir)

# # quantization config
# # Load model with 4-bit quantization if applicable (use 4-bit integer instead of 32b floats) --> reduce the required VRAM for model application
# # see, https://huggingface.co/docs/transformers/quantization
# bnb_config = BitsAndBytesConfig(
#     load_in_4bit=True,
#     bnb_4bit_use_double_quant=True,
#     bnb_4bit_quant_type="nf4",
#     bnb_4bit_compute_dtype=torch.float16,
# )



# # Model and Tokenizer initialization
# print("Model directory not found. Downloading model...")
# os.makedirs(model_dir, exist_ok=True)

# device = transformers.infer_device()
# print(f"Using device: {device}")
# model = AutoModelForCausalLM.from_pretrained(
#     model_name,
#     dtype="auto",
#     attn_implementation="flash_attention_2",  # use with 4-bit quantization,
#     # --> flash attention enables to use much larger sequence lengths without running into OOM issues
#     quantization_config=bnb_config,
# )
# model.save_pretrained(model_dir)

In [8]:
# init class for decoder and tokenizer


class DecoderModel:

    def __init__(self, model_name: str ="meta-llama/Llama-2-7b-chat-hf"):
        
        try: 
            login(token=os.getenv("HUGGINGFACE_TOKEN"))   # notebook_login
        except:
            login(token=os.environ.get("HUGGINGFACE_TOKEN"))  # former HF_TOKEN

        base_dir =  s.HF_HOME_DIR   # use default dir in .cache/
        model_dir = base_dir # / f"models--{model_name.replace("/", "--")}"  # is already .._mirror/hub/
        
        print(model_dir)

        # quantization config
        # Load model with 4-bit quantization if applicable (use 4-bit integer instead of 32b floats) --> reduce the required VRAM for model application
        # see, https://huggingface.co/docs/transformers/quantization
        bnb_config = BitsAndBytesConfig(
            load_in_4bit=True,
            bnb_4bit_use_double_quant=True,
            bnb_4bit_quant_type="nf4",
            bnb_4bit_compute_dtype=torch.float16,
        )

        self.pipeline, self.tokenizer = self.initialize_model(
            model_name, model_dir, bnb_config
        )
        

    def initialize_model(self, model_name: str, model_dir: str = None, bnb_config=None):
        
        # Model and Tokenizer initialization
        if not os.path.exists(model_dir):
            print("Model directory not found. Downloading model...")
            os.makedirs(model_dir, exist_ok=True)

            device = transformers.infer_device()
            print(f"Using device: {device}")
            model = AutoModelForCausalLM.from_pretrained(
                model_name,
                dtype="auto",
                attn_implementation="flash_attention_2",  # use with 4-bit quantization,
                # --> flash attention enables to use much larger sequence lengths without running into OOM issues
                quantization_config=bnb_config,
                # max_memory={0: "2GB", 1: "10GB"},  # distribute memory across GPUs
            )
            model.save_pretrained(model_dir)
            tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=True)
            tokenizer.save_pretrained(model_dir)

            print("Downloaded model and tokenizer")

        else:
            print(f"Using locally saved model from {model_dir}")

            model = AutoModelForCausalLM.from_pretrained(
                model_name,
                cache_dir=model_dir,
                local_files_only=True,  # tp_plan="auto" # set tensor parallel model (ie. splits model on multiple GPU)
                dtype="auto",
                attn_implementation="flash_attention_2",  # use with 4-bit quantization,
                # --> flash attention enables to use much larger sequence lengths without running into OOM issues
                quantization_config=bnb_config,
                # tp_plan="auto",  # automatically use a tensor parallelism plan based on predefined configuration of the model (i.e. partition model on both GPUs)
            )
            # print("Tensor parallel plan:", model._tp_plan)

            tokenizer = AutoTokenizer.from_pretrained(
                model_name,
                use_fast=True,
                cache_dir=model_dir,  # use fast Rust-based tokenizer, when possible
            )

        # reduce further memory usage
        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        model = model.to(device)
        model.use_checkpointing = True

        torch.cuda.empty_cache()
        torch.no_grad()

        # Pipeline setup for question answering
        pipeline = transformers.pipeline(  # load model locally from wsl .cache\
            "text-generation",
            model=model,
            tokenizer=tokenizer,
            max_new_tokens=1024, # high max token otherwise output is truncated
            device_map="auto",
        )
        return pipeline, tokenizer

    def generate_response(
        self, question: str, context: list, chunk_id: int
    ):
    
        rendered_prompt = template.render(
            context=context,  # includes also df_ci_geo info
            question=question,
        )

        print(f"Generating response for chunk_id: {chunk_id} ...")

        sequences = self.pipeline(
            rendered_prompt,  # jinja template
            max_new_tokens=1024, # use default to not truncate the LLM response
            do_sample=True,
            num_beams=1,  # select token based on probability distribution over entire model’s vocabulary
            # top_k=10,
            # top_p=0.5,
            temperature=0.2,
            # num_return_sequences=1,
            eos_token_id=self.tokenizer.eos_token_id,
            return_full_text=False,  # allow bullet point answers
        )
        # Extracting and returning the generated text
        return sequences

### Apply Llama on chunks

In [44]:
## Settings

model_name = "meta-llama/Llama-2-7b-chat-hf"
# test for HPC: "meta-llama/Llama-3.1-8B-Instruct"

df_responses = pd.DataFrame(
    columns=[
        "citation_id",
        "chunk_id",
        "infrastructure_type",
        "damage",
        "location",
        "chunk text"
    ]
)

## init LLM pipeline
decoder_model = DecoderModel(model_name=model_name)

## init output dataframe
df_responses_all = pd.DataFrame()
sentence_root_list = [] # for visualization of sentence roots and their dependencies

test_mode = True
if test_mode:
    search_path = docs_list_sample
    print("Test mode is ON. Using only a small sample of documents for testing.")
else:
    search_path = glob(str(Path(PARSED_TEXT_DIR, "*cleaned.md")))


## Start CI impact extraction
for i, filename in enumerate(search_path):

    src_language_nonengl = None 

    no_documents = len(search_path)
    filepath = Path(filename)
    filename_stem = filepath.stem



    ## load doc
    loader = DoclingLoader(filepath)  # use chunks from Docling.Loader
    doc = loader.load()


    print(f"\n\n ######## -------- Processing document [{i+1}/{no_documents}]: {filepath.name} -------- ######## \n")

    ## extract authors, publication year and title 
    author, year, title = dc.extract_citation_info(filename_stem)
    citation = f"{author} {year}".replace("  ", " ").strip()
    title = title.replace(" - ", "").replace("_cleaned", "").strip()


    # check language of document based on document title
    # check title rather than chunk_text as later might contain also footer or header info in another languages
    src_language = langdetect.detect(title)

    if src_language != "en":
        supported_languages = ["fr", "de", "es", "it", "itc", "nl"]
        src_language_nonengl = src_language
        if src_language not in supported_languages:
            raise ValueError(f"Unsupported source language: {src_language}")



    print(f"\n ######## -------- Getting geolocation of CI assets -------- ######## \n")
    ## Create New entity for transport infrastructure and apply it on any doc

    ## TODO make as pydantic class with fixed attributes
    df_ci_geo = pd.DataFrame(
        columns=[
            "citation_id",
            "chunk_id",
            "ci_entity",
            "geo_entity",
            "chunk text",            
        ]
    )

    print(f"\n ######## -------- Relationship extraction: CI-GEO pairs -------- ######## \n")

    for chunk_no, chunk in enumerate(doc):

        if src_language_nonengl:

            print(f"\n ######## -------- Translating {langdetect.detect(chunk.page_content)} text to `en` -------- ######## \n")
            chunk.page_content = tm.translate_2_english(chunk.page_content)



        ## apply customized NLP pipeline -> returns same chunk.text but with CI_entities annotations and without linebreak symbols
        for nlp_chunk in nlp.pipe([chunk.page_content]):

            ci_entities = [ent for ent in nlp_chunk.ents if ent.label_ in ["CI_TYPE"]]

            # check if chunk has no CI_TYPE entities
            if not ci_entities:
                print(f"Chunk [{chunk_no}], No CI_TYPE entities found {len(ci_entities)}")
                # print(f"\nChunk [{chunk_no}], No. CI_TYPE and FAC entities: {len(ci_entities)}")
                continue
      

            print(f"Chunk [{chunk_no}], {len(ci_entities)} CI_TYPE  entities found")
            

            for ci_ent in ci_entities:
                print("Ci_ent:", ci_ent.text)

                # Get the root of the sentence 
                sentence_root = ci_ent.sent.root

                if sentence_root.dep_ in ("ROOT", "dobj", "conj"):  # TODO probably enough when set only to ROOT 
                    # sentence_root_list.append(sentence_root.dep_) # for visualization of sentence roots and their dependencies
                    head = ci_ent.root.head
                    # print(f" `{sentence_root, sentence_root.dep_}` is sentence_root, \n `{head}` the direct head of ci entity `{ci_ent}`")

                    subjs = list(sentence_root.lefts) + list(sentence_root.rights)
                    for subj in subjs:
                        for descendant in subj.subtree:

                            # check if descendant is a location entity
                            if descendant.ent_type_ in ("GPE", "LOC"):
                                descendant_loc = descendant
                                print("  ", ci_ent.text, "->", descendant_loc.text)
                                # assert subj is descendant_loc or subj.is_ancestor(descendant_loc)

                                ## Special case handling - one root but multiple ci-geo pairs
                                try:
                                    first_ci_ancestor = [ancestor.text for ancestor in descendant_loc.ancestors if ancestor.ent_type_ in ("CI_TYPE", "FAC")][0]
                                    if ci_ent.text == first_ci_ancestor:
                                        print(
                                            "descendant of location:", descendant_loc.text, descendant_loc.dep_, 
                                            "\nits ancestors:", [ancestor.text for ancestor in descendant_loc.ancestors] # all parents up to root.head
                                        )
                                        print("  ", ci_ent.text, "-->", descendant_loc.text)
                                        
                                        ## write as dict entry incl chunk_id, ci_entity, geo_entity, distance
                                        result_dict = {
                                            "citation_id": citation,
                                            "chunk_id": chunk_no,
                                            "ci_entity": ci_ent.text,
                                            "geo_entity": descendant_loc.text,
                                            "case_type": "special",
                                            "chunk text": nlp_chunk,
                                        }
                                        df_ci_geo = pd.concat(
                                            [df_ci_geo, pd.DataFrame([result_dict])], ignore_index=True
                                        )

                                ## usual case handling
                                except IndexError:
                                    ## write as dict entry incl chunk_id, ci_entity, geo_entity, distance
                                    result_dict = {
                                        "citation_id": citation,
                                        "chunk_id": chunk_no,
                                        "ci_entity": ci_ent.text,
                                        "geo_entity": descendant_loc.text,
                                        "case_type": "usual",
                                        "chunk text": nlp_chunk,
                                    }
                                    df_ci_geo = pd.concat(
                                        [df_ci_geo, pd.DataFrame([result_dict])], ignore_index=True
                                    )

                else:
                    # print("\nNo CI_TYPE or FAC entities found in this chunk.")
                    continue

    ## post-process of DF CI-GEO pairs
    df_ci_geo = df_ci_geo.drop_duplicates(
        subset=["citation_id", "chunk_id", "ci_entity", "geo_entity","case_type", "chunk text"]
        )# .reset_index(drop=True, inplace=True)


    print(f"\n  #############  -------- Text-2-Data: {filepath.name} -------- #############  \n")

    ## apply decoder on each chunk in document
    ## TODO replace iteration by loading entire document and use recursive chunking from langchain
    for j, chunk in enumerate(doc):

        if df_ci_geo.loc[df_ci_geo["chunk_id"] == j].empty:
            continue

        context = [
            {
                "text": chunk.page_content,
                "citation": citation,
                "title": filename_stem, 
                # TODO check if chunk id is unique per doc and if multiple ci_loc pairs eixist (then just tke first one)
                "ci_locations": df_ci_geo.loc[df_ci_geo["chunk_id"] == j],
            },
        ]

        # apply LLM
        response = decoder_model.generate_response(
            question=question_1, context=context, chunk_id=j
        )

        ## postprocess response
        resp = response[0]["generated_text"].replace("\n", "")
        try:
            resp = (resp.split("]")[0] + "]")  # remove potential text outside of json object
            df_resp = pd.read_json(StringIO(resp))
            df_resp["citation_id"] = citation  # add citation info
            df_resp["chunk_id"] = j  # add chunk id as identifier
            df_resp["chunk text"] = chunk.page_content
            df_responses = pd.concat([df_responses, df_resp], ignore_index=True)
        except ValueError as e:
            print(f"Cannot add response to output dataframe: {e}, \n{resp}")


    df_responses_all = pd.concat([df_responses_all, df_responses], ignore_index=True)


    # clean up after each document
    gc.collect()
    torch.cuda.empty_cache()  # mainyl after training needed, small effect when LLM applied only for infernece
    torch.no_grad()

Note: Environment variable`HF_TOKEN` is set and is the current active token independently from the token you've just configured.


/home/a-buch/Documents/TUB_DWN/_PROJECTS/CI-impacts-information-retrieval/notebooks/huggingface_mirror/hub
Using locally saved model from /home/a-buch/Documents/TUB_DWN/_PROJECTS/CI-impacts-information-retrieval/notebooks/huggingface_mirror/hub


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

OutOfMemoryError: CUDA out of memory. Tried to allocate 86.00 MiB. GPU 0 has a total capacity of 5.67 GiB of which 64.81 MiB is free. Process 3574 has 112.18 MiB memory in use. Including non-PyTorch memory, this process has 5.29 GiB memory in use. Of the allocated memory 4.71 GiB is allocated by PyTorch, and 470.27 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

a-buch-ThinkPad-X1-Extreme-Gen-4i


#### response

In [ ]:
safety_df = df_responses_all.copy()

# save LLM 1 output to disk along with prompt text


if not os.path.isfile(OUTPUT_LLM1_FILEPATH):

    print(f"Saving prompt and LLM response {OUTPUT_LLM1_FILEPATH.stem} [.txt, .csv] to {OUTPUT_LLM1_FILEPATH.parent} ...")
    with open(OUTPUT_PROMPT_FILEPATH, "w") as f:
        f.write(
            template.render(context=context, question=question_1)
        )
    df_responses_all.to_csv(OUTPUT_LLM1_FILEPATH, index=False)
else:
    print(f"Output file {Path(OUTPUT_LLM1_FILEPATH).stem} already exists. Skip saving to avoid overwriting ...")


    # # check if file exists already
    # if os.path.exists(OUTPUT_LMM1_FILEPATH):
    #     print(f"File {OUTPUT_LMM1_FILEPATH.name} already exists. Not saving to disk.")
    #     pass
    # else:
    #     print(f"Saving LLM 1 output to {OUTPUT_LMM1_FILEPATH}")
    #     df_responses_all.to_csv(OUTPUT_LMM1_FILEPATH, index=False)

    




## Evaluation - only manually

### Manual comparison CI_location_table vs LLm response


#### chunk 5
In Germany, road and railway infrastructure was severely damaged as documented exemplarily in Fig. 1. Cost estimates reach up to EURO 2 billion Euro (MDR, 2021). More than 130 km of motorways were closed directly after the event, of which 50 km were still closed two months later, with an estimated repair cost of EUR 100 million (Hauser, 2021). Of the 112 bridges in the flooded 40 km of the Ahr valley (Rhineland-Palatinate), 62 bridges were destroyed, 13 were severely damaged and only 35 were in operation a month after the flood event (MDR, 2021). Over 74 km of roads, paths and bridges in the Ahr valley have been (critically) damaged. In some cases, repairs are expected to take months to years (Zeit Online, 2021). For example, major freeway sections, including parts of the A1 motorway, were closed until early 2022 (24Rhein, 2022). In addition, about 50 000 cars were damaged, causing insurance claims of some EUR 450 million (ADAC, 2021). The German railway provider Deutsche Bahn expects asset damages of around EUR 1.3 billion. Among other things, 180 level crossings, almost 40 signal'

In [ ]:
doc[5].page_content

In [ ]:
df_ci_geo[df_ci_geo["chunk_id"] == 5]

In [ ]:
df_responses[df_responses["chunk_id"] == 5]

In [ ]:
print(
    f"CI_TYPE \n LLM responses:\n {df_responses.infrastructure_type.unique()}, \n\n NER pairs:\n {df_ci_geo.ci_entity.unique()}"
)

In [ ]:
print(
    f"LOCATION:\n LLm responses:\n {df_responses.location.unique()}, \n\n NER pairs:\n {df_ci_geo.geo_entity.unique()}"
)

### chunk 13 +14
'We found no information regarding direct impact on solid-waste facilities as a result of the flood event. However, there is a large pressure on the solid-waste sector to clean the affected areas; 1 month after the event, we observed dozens of large temporary waste fills and frequent incidences of oil pollution in Rhineland-Palatinate during a field visit. In the Ahrweiler district alone, the flood caused as much solid waste as normally would be collected over 30 years. In Belgium, the amount of solid waste is estimated around 160 000 t, stored at several places, such as the abandoned highway track A601. This highway has been used for approximately 9 months as a temporary storage for debris (Couplez, 2022). In the Netherlands, there have been primarily problems with waste deposits along the river banks, which is mostly the solid waste transported by the river from further upstream. Thousands of tonnes of tree debris (logs and\nNat. Hazards Earth Syst. Sci., 22, 3831–3838, 2022\nE. E. Koks et al.: Flood impacts to infrastructure'


'of running water and electricity (Ärzte Zeitung, 2021). After 1.5 months, medical care was guaranteed again in the most affected regions in Rhineland-Palatinate (Hochwasser Ahr, 2021c). In the state of North Rhine-Westphalia, approximately 68 hospitals have been affected, of which several have been affected severely and will take at least 1.5 years to be rebuilt (Fig. 2). Direct damages are estimated to be at least EUR 100 million to repair all medical facilities (Korzilius, 2021). In the town of Eschweiler (Germany), for example, the basement of the hospital was flooded, as well as the outbuildings and the entire outdoor area. The power supply collapsed, the entire building technology was destroyed and some 300 patients had to be evacuated by helicopter. Property damage is expected to be around EUR 50 million. Within 3.5 weeks, the hospital was partly operational, and within 3 months, all hospital operations continued normally (SAH Eschweiler, 2021). The Mutterhaus Ehrang hospital in Trier (Germany) is now permanently closed as the hospital is too severely damaged to rebuild. Furthermore, in the region of Rhineland-Palatinate (Germany), 19'


In [ ]:
doc[14].page_content

In [ ]:
df_ci_geo[df_ci_geo["chunk_id"].isin([13, 14])]

In [ ]:
df_responses[df_responses["chunk_id"].isin([13, 14])]

In [ ]:
print(
    f"CI_TYPE \n LLM responses:\n {df_responses.infrastructure_type.unique()}, \n\n NER pairs:\n {df_ci_geo.ci_entity.unique()}"
)

In [ ]:
print(
    f"LOCATION:\n LLm responses:\n {df_responses.location.unique()}, \n\n NER pairs:\n {df_ci_geo.geo_entity.unique()}"
)

## Test alternative approaches for entity linking / relation extraction

In [ ]:
import torch
import gc

print(torch.cuda.memory_summary(device=None, abbreviated=False))
# # empyty CUDA cache
gc.collect()

torch.cuda.empty_cache()
torch.no_grad()
# print(torch.cuda.memory_summary(device=None, abbreviated=False))